lib/dialogs/rule.js

```diff
@@ -96,34 +99,35 @@ async function completeProgram(dlg, program, skipConfirmation) {
         }
     }
 
-    let ok = await slotFillProgram(dlg, program);
-    if (!ok)
-        return { ok };
-
-    let icon = null;
+    let icon;
     for (let [, prim] of program.iteratePrimitives()) {
         if (prim.selector.isBuiltin)
             continue;
         let newIcon = Helpers.getIcon(prim);
         if (newIcon)
             icon = newIcon;
     }
+    dlg.icon = icon;
+
+    let hasSlots = false;
+    for (let [schema, slot] of program.iterateSlots()) {
+        if (slot instanceof ThingTalk.Ast.InputParam ) {
```

Comment:
Does it matter if the slot is already filled here? What about `$?` in filters?

Reply:
`iterateSlots` returns all the slots in the program including the ones in filters, right?